From 941098459db6556b837194f40c076b08d51137cb Mon Sep 17 00:00:00 2001
From: "Alex K." <8418476+fearful-symmetry@users.noreply.github.com>
Date: Tue, 20 Jan 2026 09:31:50 -0800
Subject: [PATCH] Add array checks to packetbeat procfs parser (#48428)

* init commit

* fix int parsing

* linter, uint16 cast

* changelog

* linter cleanup

* add test

* remove old IO lib
---
 ...1768584444-fix-procfs-network-parsers.yaml | 45 +++++++++++++++++++
 packetbeat/procs/procs_linux.go               | 19 +++++---
 packetbeat/procs/procs_linux_test.go          | 13 +++---
 3 files changed, 65 insertions(+), 12 deletions(-)
 create mode 100644 changelog/fragments/1768584444-fix-procfs-network-parsers.yaml

diff --git a/changelog/fragments/1768584444-fix-procfs-network-parsers.yaml b/changelog/fragments/1768584444-fix-procfs-network-parsers.yaml
new file mode 100644
index 000000000000..0971deb75e17
--- /dev/null
+++ b/changelog/fragments/1768584444-fix-procfs-network-parsers.yaml
@@ -0,0 +1,45 @@
+# REQUIRED
+# Kind can be one of:
+# - breaking-change: a change to previously-documented behavior
+# - deprecation: functionality that is being removed in a later release
+# - bug-fix: fixes a problem in a previous version
+# - enhancement: extends functionality but does not break or fix existing behavior
+# - feature: new functionality
+# - known-issue: problems that we are aware of in a given version
+# - security: impacts on the security of a product or a user’s deployment.
+# - upgrade: important information for someone upgrading from a prior version
+# - other: does not fit into any of the other categories
+kind: bug-fix
+
+# REQUIRED for all kinds
+# Change summary; a 80ish characters long description of the change.
+summary: fix procfs network parsers
+
+# REQUIRED for breaking-change, deprecation, known-issue
+# Long description; in case the summary is not enough to describe the change
+# this field accommodate a description without length limits.
+# description:
+
+# REQUIRED for breaking-change, deprecation, known-issue
+# impact:
+
+# REQUIRED for breaking-change, deprecation, known-issue
+# action:
+
+# REQUIRED for all kinds
+# Affected component; usually one of "elastic-agent", "fleet-server", "filebeat", "metricbeat", "auditbeat", "all", etc.
+component: packetbeat
+
+# AUTOMATED
+# OPTIONAL to manually add other PR URLs
+# PR URL: A link the PR that added the changeset.
+# If not present is automatically filled by the tooling finding the PR where this changelog fragment has been added.
+# NOTE: the tooling supports backports, so it's able to fill the original PR number instead of the backport PR number.
+# Please provide it if you are adding a fragment for a different PR.
+# pr: https://github.com/owner/repo/1234
+
+# AUTOMATED
+# OPTIONAL to manually add other issue URLs
+# Issue URL; optional; the GitHub issue related to this changeset (either closes or is part of).
+# If not present is automatically filled by the tooling with the issue linked to the PR number.
+# issue: https://github.com/owner/repo/1234
diff --git a/packetbeat/procs/procs_linux.go b/packetbeat/procs/procs_linux.go
index 8ad952a1e26e..a5ce78bac338 100644
--- a/packetbeat/procs/procs_linux.go
+++ b/packetbeat/procs/procs_linux.go
@@ -141,13 +141,13 @@ func findSocketsOfPid(prefix string, pid int) (inodes []uint64, err error) {
 		}
 
 		if strings.HasPrefix(link, "socket:[") {
-			inode, err := strconv.ParseInt(link[8:len(link)-1], 10, 64)
+			inode, err := strconv.ParseUint(link[8:len(link)-1], 10, 64)
 			if err != nil {
 				logp.Debug("procs", "%s", err.Error())
 				continue
 			}
 
-			inodes = append(inodes, uint64(inode))
+			inodes = append(inodes, inode)
 		}
 	}
 
@@ -204,10 +204,10 @@ func parseProcNetProto(input io.Reader, ipv6 bool) ([]*socketInfo, error) {
 			continue
 		}
 
-		uid, _ := strconv.Atoi(string(words[7]))
+		uid, _ := strconv.ParseUint(string(words[7]), 10, 32)
 		sock.uid = uint32(uid)
-		inode, _ := strconv.Atoi(string(words[9]))
-		sock.inode = uint64(inode)
+		inode, _ := strconv.ParseUint(string(words[9]), 10, 64)
+		sock.inode = inode
 
 		sockets = append(sockets, &sock)
 	}
@@ -230,7 +230,7 @@ func hexToIPPort(str []byte, ipv6 bool) (net.IP, uint16, error) {
 		return nil, 0, err
 	}
 
-	port, err := strconv.ParseInt(string(words[1]), 16, 32)
+	port, err := strconv.ParseUint(string(words[1]), 16, 16)
 	if err != nil {
 		return nil, 0, err
 	}
@@ -255,8 +255,13 @@ func hexToIpv4(word string) (net.IP, error) {
 
 func hexToIpv6(word string) (net.IP, error) {
 	p := make(net.IP, net.IPv6len)
+	if len(word) < 32 {
+		return nil, fmt.Errorf("got ip6 address of invalid length: %d", len(word))
+	}
 	for i := 0; i < 4; i++ {
-		part, err := strconv.ParseUint(word[i*8:(i+1)*8], 16, 32)
+		start := i * 8
+		end := (i + 1) * 8
+		part, err := strconv.ParseUint(word[start:end], 16, 32)
 		if err != nil {
 			return nil, err
 		}
diff --git a/packetbeat/procs/procs_linux_test.go b/packetbeat/procs/procs_linux_test.go
index 3923bc21109a..d2ce507c66ca 100644
--- a/packetbeat/procs/procs_linux_test.go
+++ b/packetbeat/procs/procs_linux_test.go
@@ -20,16 +20,14 @@
 package procs
 
 import (
-	"io/ioutil"
 	"os"
 	"path/filepath"
 	"testing"
 
-	"github.com/elastic/elastic-agent-libs/logp"
+	"github.com/stretchr/testify/require"
 )
 
 func TestFindSocketsOfPid(t *testing.T) {
-	logp.TestingSetup()
 
 	proc := []testProcFile{
 		{path: "/proc/766/fd/0", isLink: true, contents: "/dev/null"},
@@ -44,7 +42,7 @@ func TestFindSocketsOfPid(t *testing.T) {
 	}
 
 	// Create fake proc file system
-	pathPrefix, err := ioutil.TempDir("", "find-sockets")
+	pathPrefix, err := os.MkdirTemp("", "find-sockets")
 	if err != nil {
 		t.Error("TempDir failed:", err)
 		return
@@ -65,6 +63,11 @@ func TestFindSocketsOfPid(t *testing.T) {
 	assertUint64ArraysAreEqual(t, []uint64{7619, 7620}, inodes)
 }
 
+func TestParseInvalidIp6Addr(t *testing.T) {
+	_, err := hexToIpv6("B80D012000000000FFFF2301EFCD")
+	require.Error(t, err)
+}
+
 func TestParse_Proc_Net_Tcp(t *testing.T) {
 	socketInfo, err := socketsFromProc("../tests/files/proc_net_tcp.txt", false)
 	if err != nil {
@@ -116,7 +119,7 @@ func createFakeDirectoryStructure(prefix string, files []testProcFile) error {
 		}
 
 		if !file.isLink {
-			err = ioutil.WriteFile(filepath.Join(prefix, file.path),
+			err = os.WriteFile(filepath.Join(prefix, file.path),
 				[]byte(file.contents), 0o644)
 			if err != nil {
 				return err
